### PR TITLE
Change MIR linker RFC to stable.

### LINKED ISSUE #1588: Tracking issue for MIR linker
## Summary

This is a tracking issue for the RFC `mir_linker` which will enable users to use non-generic + non-inlined methods of the standard library.

During development, this feature will be gated by the unstable option `--mir-linker`.

## Major Steps

Suggestion of major steps:

- [x] Get feedback on the RFC PR.
- [x] Implement compiler options `--reachability=[legacy | none | harnesses]`
- [x] Create a mechanism to generate and ship Kani sysroot.
- [x] Introduce unstable user option `--mir-linker`.
- [x] Implement compiler option `--reachability=pub_fns`.
- [ ] Stabilization PR.

* Broken down issues will be tracked by https://github.com/model-checking/kani/milestone/12

## Open Questions

- Should we build or download the sysroot during `kani setup`?

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Update the RFC to capture the current state of the MIR linker and mark it as Stable.

### Resolved issues:

Resolves #1588 

### Related RFC:

Optional #1588.

### Testing:

* How is this change tested? N/A

* Is this a refactor change? Nope

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: rfc/src/rfcs/0001-mir-linker.md [hunk 1]
```diff
@@ -97,7 +97,7 @@ The two options to be included in this RFC is starting from all local harnesses
 The `kani-compiler` behavior will be customizable via a new flag:
 
   ```
-  --reachability=[ harnesses | pub_fns |  none | legacy ]
+  --reachability=[ harnesses | pub_fns |  none | legacy | tests ]
```
Comment (zhassan-aws) [inline rfc/src/rfcs/0001-mir-linker.md:100]: Should we remove `legacy`?

File: rfc/src/rfcs/0001-mir-linker.md [hunk 2]
```diff
@@ -108,11 +108,12 @@ where:
  reachability analysis. No goto-program will be generated.
   This will be used to compile dependencies up to the MIR level.
    `kani-compiler` will still generate artifacts with the crate's MIR.
+ - `tests`: Use the functions marked as tests with `#[tests]` as the starting points for the analysis.
  - `legacy`: Keep `kani-compiler` current behavior by using
```
Comment (zhassan-aws) [inline rfc/src/rfcs/0001-mir-linker.md:112]: Same here.

File: rfc/src/rfcs/0001-mir-linker.md (hunk 1 above)
Comment (celinval) [reply rfc/src/rfcs/0001-mir-linker.md:100]: We still have `legacy` which is used internally to test code generation of the standard library.

File: rfc/src/rfcs/0001-mir-linker.md (hunk 2 above)
Comment (celinval) [reply rfc/src/rfcs/0001-mir-linker.md:112]: I could update this to say that this is the behavior similar to how `rustc` compile crates into separate objects.

File: rfc/src/rfcs/0001-mir-linker.md (hunk 1 above)
Comment (zhassan-aws) [reply rfc/src/rfcs/0001-mir-linker.md:100]: OK.

File: rfc/src/rfcs/0001-mir-linker.md (hunk 2 above)
Comment (zhassan-aws) [reply rfc/src/rfcs/0001-mir-linker.md:112]: That would be useful.

**Review by zhassan-aws — APPROVED** [review]

File: rfc/src/rfcs/0001-mir-linker.md (hunk 2 above)
Comment (celinval) [reply rfc/src/rfcs/0001-mir-linker.md]: Done. Let me know what you think.

Comment (zhassan-aws) [reply rfc/src/rfcs/0001-mir-linker.md]: Looks good. Thanks.

File: rfc/src/rfcs/0001-mir-linker.md [hunk 3]
```diff
@@ -108,11 +108,12 @@ where:
  reachability analysis. No goto-program will be generated.
   This will be used to compile dependencies up to the MIR level.
    `kani-compiler` will still generate artifacts with the crate's MIR.
- - `legacy`: Keep `kani-compiler` current behavior by using
-   `rustc_monomorphizer::collect_and_partition_mono_items()` which respects the crate boundary.
-   This will generate a goto-program for each crate compiled by `kani-compiler`, and it will still have the same
-   `std` linking issues.
-   *This option will be removed as part of the `rfc` stabilization.*
+ - `tests`: Use the functions marked as tests with `#[tests]` as the starting points for the analysis.
```
Comment (danielsn) [inline rfc/src/rfcs/0001-mir-linker.md:111]: Do we also include functions marked `proof`?

File: rfc/src/rfcs/0001-mir-linker.md [hunk 4]
```diff
@@ -203,7 +204,7 @@ See the table bellow for the breakdown of time (in seconds) taken for each major
 the harness verification:
 
 
-| Stage                     | MIR Linker | Alternative |
+| Stage                     | MIR Linker | Alternative 1 |
```
Comment (danielsn) [inline rfc/src/rfcs/0001-mir-linker.md:207]: No alternative 2 numbers?

**Review by danielsn — APPROVED** [review]
Two small questions, but otherwise lgtm

File: rfc/src/rfcs/0001-mir-linker.md (hunk 3 above)
Comment (celinval) [reply rfc/src/rfcs/0001-mir-linker.md:111]: Not for the `tests` mode.

File: rfc/src/rfcs/0001-mir-linker.md (hunk 4 above)
Comment (celinval) [reply rfc/src/rfcs/0001-mir-linker.md:207]: No, I could try to collect from my old branches, but I don't think it's worth it at this point. If I remember correctly, for Alternative 2, the compilation time was closer to the MIR Linker, but everything else was closer to Alternative 1.

**Pull request merged**